### ISSUE TITLE
How do I know when tap is done?

### ISSUE THREAD

**Issue by jtlapp** [tapjs/tapjs#288]
UPDATE: The answer to the question is this: Call `tap.end()` after you've loaded all the tests via `tap.test()`, and then catch completion in the callback for `tap.tearDown(cb)`.

I'm running tap in a forked process. I have to call `process.exit(0)` because a node-forked process won't terminate on its own. But I can't figure out when to do this. Neither the 'complete' event nor the 'end' event is called to report completion. My `setImmediate()` function is called before most of the tests run. I'm piping the tap output to my own Writable, but I can't figure out how to detect completion even without piping. (And `end()` is not called on my Writable.)

How do I learn that tap is done so I can call `process.exit()`? Thanks for your help!

**Comment by isaacs** [comment]
Tap should definitely exit on its own when it's done.  Can you share your test script?

**Comment by jtlapp** [comment]
HI @isaacs. I don't have a `node-tap` test script for you, but you can look over [the code in my 'fork-tests' branch](https://github.com/jtlapp/subtap/tree/fork-tests/bin). I've been exploring variations without success. `subtap.js` forks `_runfile.js`. These are the two relevant files.

**Comment by jtlapp** [comment]
I thought to write a unit test for you but realized that I have no idea what _should_ signal completion. Which of the following should it be? Or what else?
- `tap.on('complete', theEnd)`
- `tap.on('end', theEnd)`
- `tap.pipe(writable).end()`
- `tap.pipe(writable).on('finish', theEnd)`
- `setImmediate(theEnd)`

I can't wait on `process.on('beforeExit')` or `process.on('exit')` because those events do not occur in a node-forked process until `process.exit()` is explicitly called.

**Comment by jtlapp** [comment]
Maybe tap needs an `idle` event to signal that no tap-related work is currently pending. Tap can't know whether another test file is about to be loaded, or whether a test will later be issued on timeout, but it should know whether it's presently waiting on anything.

Is there any way for me to get the size of the next-tick queue?

**Comment by jtlapp** [comment]
Well, I've monkey-patched `tap` to count outstanding tests and exit when all have ended. I don't know how generally it works, because I reverse-engineered more from behavior than from source. (I have poor eyesight at 47 and can't track two-space indentations well.) I've tested it for explicit calls to `t.end()`, for deferred plans, and for returned promises. [Here's the relevant code](https://github.com/jtlapp/subtap/blob/fork-tests/bin/_runfile.js#L26).

**Comment by jtlapp** [comment]
Drats. I think I see a much easier way -- watching the TAP output. `tap` knows when it's hit the end and says so. I just have to look for that mark in the output.

**Comment by isaacs** [comment]
So, before you go reinventing wheels, it seems like this project is reproducing a lot of stuff that's already in node-tap, in the sense that node-tap already has a bin that spawns child processes and can present output as nested tests.  (Try running `tap test/*.js --reporter=tap`.)  Instead of using ICP to communicate bailout and completion, it just looks at the tap output.  There's even a builtin function, `tap.spawn`, which will run a child test as a sub-process and handle its output.

If your goal is to run subtests that have ICP capabilities, you can do `tap.spawn(process.execFile, [theTestFile], { stdio: [ 'pipe', 'pipe', 'pipe', 'ipc' ] })`, but I don't know offhand if there's a way to get at the child process object.  It wouldn't be hard to add support for that, maybe a callback that gets the child process object as an argument once it's set up and you're at that point in the test set.

If you are dead set on going your own way (which is totally fine, allowed, and can be a source of much goodness), then take a look at what the code in tap's `bin/run.js` is doing.

The reason that tap's bin runs tests as a standalone child proc, rather than using `require()` is that it ensures that the test can be run identically with the bin runner or by typing `node test.js`, with a "fresh" environment where the test itself is the `require.main` entry point.  This also obviates the need for a "when to close" indication, because Node closes when the event loop is empty, and the exit handler in `lib/root.js` prints out the final stuff.

One option, if you're going to load the test into the current process, is to use `tap.tearDown(someFunction)`, which will require tap to exit once there's nothing pending.  (Otherwise it would be impossible to do things like shut down a server at the end of your test run.)  So you could just have your child proc do `tap.tearDown(function () { process.exit() })`.  No monkey-patching necessary, it does exactly what you're describing here, and is supported and tested already :)  http://www.node-tap.org/api/#tteardownfunction

**Comment by jtlapp** [comment]
Thanks @isaacs. I already have a simple solution in the works. I need IPC because I'm communicating more than bailout and completion. My alternative was to add TAP comments to the stream and read data from them.

I was also making a mistake by not calling `tap.end()` when I knew I wouldn't be queuing any more tests. If I do that, I get tap 'complete' and 'end' events, though I'm yet sure of the difference.

`tap.tearDown()` and the various emitted events don't seem to be documented outside the source, and I'm having trouble reading indentation in the source, so I'm kinda going blind here.

Thanks for your help. It looks like I now have several solutions.

**Comment by isaacs** [comment]
Yes, calling `tap.end()` explicitly is also an option.

The public API (including the tearDown function) is documented at http://www.node-tap.org/api/

What are you finding is missing from the docs that would be helpful?  I don't have docs for all emitted events, because that is an internal API that isn't really supported publicly.  Apart from the standard Node.js [readable stream interface](https://nodejs.org/api/stream.html#stream_readable_streams), it only emits `result`, `complete`, and `bailout` events, which are likely what you'd expect.

As far as the indentation, I use the [standard](http://npm.im/standard) linter on this project, which is very popular, so it is probably worthwhile to get familiar with it, since you'll surely encounter it a lot.

**Comment by jtlapp** [comment]
Woo! [It works](https://github.com/jtlapp/subtap/blob/fork-tests/bin/_runfile.js), no monkey patching, and it's simple. Thank you!

Looking through the docs now, I see that all the information I needed was there. It's hard to retrospectively know what threw me, but I'm guessing:
- Virtually no examples show `.end()` being called on `tap` itself, even though it makes sense because the docs clearly state that the module returns a `Test` object.
- The `tearDown()` docs say "Run the supplied function when t.end() is called", which (as stated) is not quite the behavior I want. I know I can't call `process.exit()` in place of my `tap.end()`. It seems that `tearDown()` instead runs when all outstanding tests have completed.

Even so, I see now that there was enough information in the docs to have suggested that I experiment with this solution. The docs are pretty darn good.

Thank you again for all the help you've given me!

**Comment by isaacs** [comment]
Awesome!  Yeah, I guess it should be clearer that it's not "when t.end() is called" but rather, "when the test ends".  (It'll also run if you use a plan, for example, or set `autoend: true`.)  I'll update the docs to correct that.

Hey, one thing that jumps out at me, is that you're doing `process.send(...)` and then immediately calling `process.exit(0)`.  I think that's a race condition, since IPC is asynchronous.  It's probably hard to see in simple cases, but if you had a lot of stuff going on in that channel, it might result in the child process exiting before the message has been fully written to the channel.  It'd be safer to have the parent process send a reply indicating that it got the message, or even just have the parent process kill the child process when the message is received.

**Comment by jtlapp** [comment]
Thanks for that tip. I should have caught that. But `tap` has an on-exit handler that kicks in after the kill signal, and that handler is not happy:

```
/Users/joe/repos/qloader/node_modules/tap/lib/test.js:321
      throw er
      ^

Error: test after end() was called
    at Test._onTimeout (/Users/joe/repos/qloader/node_modules/tap/lib/test.js:177:5)
    at EventEmitter.<anonymous> (/Users/joe/repos/qloader/node_modules/tap/lib/root.js:125:7)
    at emitTwo (events.js:106:13)
    at EventEmitter.emit (events.js:191:7)
    at emit (/Users/joe/repos/qloader/node_modules/signal-exit/index.js:68:11)
    at process.listener (/Users/joe/repos/qloader/node_modules/signal-exit/index.js:82:7)
    at emitNone (events.js:86:13)
    at process.emit (events.js:185:7)
    at processEmit (/Users/joe/repos/qloader/node_modules/signal-exit/index.js:146:32)
    at Signal.wrap.onsignal (internal/process.js:201:44)
```

**Comment by isaacs** [comment]
So kill it with SIGKILL, rather than SIGTERM.  (SIGTERM is for timeouts, so tap reads that as a test timing out, but still, it looks like you may have found a bug.)

**Comment by jtlapp** [comment]
Thank you. `SIGKILL` did the trick.

**Comment by jtlapp** [comment]
My new tool [`subtap`](https://github.com/jtlapp/subtap) breaks going from tap@6.3.2 to tap@7.0.0 . It appears that root-level assertions run in my fork but that subtests aren't allowed to run. It's probably something I'm doing inappropriately and I'll continue to look at it.

**Comment by jtlapp** [comment]
This is the only code that seems relevant to the problem, in root.js:

``` js
  if (!tap._ended) {
    tap._onTimeout(extra)
  } else {
    ....
  }
```

This was just:

``` js
tap._onTimeout(extra)
```

UPDATE: Nope. I tried restoring just this, but it didn't help. Looking elsewhere...

**Comment by jtlapp** [comment]
My first real clue: the TAP output under version 7.0.0 indents root-level subtest names with four spaces, while before version 7.0.0, these names are not indented. When I run the test file directly with node, the indentation is correct. The problem has something to do with forking or my monkey patch.

**Comment by jtlapp** [comment]
CORRECTION! I had that backwards. `tap` 7.0.0 is outputting subtest names on the left margin. `tap-parser` does not like that. It has nothing to do with my software.

**Comment by jtlapp** [comment]
Problem solved. You also updated `tap-parser`. If I update `tap-parser` too, my tool works with 7.0.0.

Unfortunately, my tool is designed to work with the user's existing installation of `tap`. In order for people to be able to use my tool, I'll have to force them to upgrade `tap` to 7.0.0. I'm not sure whether that's a reasonable requirement for a test runner.

**Comment by isaacs** [comment]
@jtlapp Just use the latest tap-parser. It properly handles both the indented and unindented child test formats, and should be 99% backwards compatible except in some very rare edge cases. It also handles "buffered" subtest formats that the Test2 cpan module uses.
